tests/test_router_direct_custom_quote_in_given_out.py

```diff
@@ -0,0 +1,87 @@
+
+        # Get the optimal quote for the given token pair
+        # Direct custom quote does not support multiple routes, so we request single/multi hop pool routes only
+        optimal_quote = ExactAmountOutQuote.run_quote_test(environment_url, token_out, denom_in, False, True, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        pool_id = ','.join(map(str, optimal_quote.get_pool_ids()))
+        denoms_in = ','.join(map(str, optimal_quote.get_token_in_denoms()))
+
+        quote = self.run_quote_test(environment_url, token_out, denoms_in, pool_id, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        # All tokens have the same default exponent, resulting in scaling factor of 1.
+        spot_price_scaling_factor = 1
+
+        # Compute expected base out quote spot price
+        # First, get the USD price of each denom, and then divide to get the expected spot price
+        in_base_usd_quote_price = conftest.get_usd_price_scaled(denom_in)
+        out_base_usd_quote_price = conftest.get_usd_price_scaled(coin.denom)
+        expected_in_base_out_quote_price = out_base_usd_quote_price / in_base_usd_quote_price 
+
+        # Compute expected token out
+        expected_token_in = int(coin.amount) * expected_in_base_out_quote_price
+
+        token_out_amount_usdc_value = in_base_usd_quote_price * coin.amount
+
+        # Chose the error tolerance based on amount in swapped.
+        error_tolerance = Quote.choose_error_tolerance(token_out_amount_usdc_value)
+
+        # Validate that price impact is present.
+        assert quote.price_impact is not None
+
+        # Validate quote results
+        ExactAmountOutQuote.validate_quote_test(quote, coin.amount, coin.denom, spot_price_scaling_factor, expected_in_base_out_quote_price, expected_token_in, denom_in, error_tolerance, True)
+
+    def run_quote_test(self, environment_url, token_out, denom_in, pool_id, expected_latency_upper_bound_ms, expected_status_code=200) -> QuoteExactAmountOutResponse:
+        """
+        Runs exact amount out test for the /router/quote endpoint with the given input parameters.
+
+        Does basic validation around response status code and latency
+
+        Returns quote for additional validation if needed by client
+
+        Validates:
+        - Response status code is as given or default 200
+        - Latency is under the given bound
+        """
+
+        sqs_service = conftest.SERVICE_MAP[environment_url]
+
+        start_time = time.time()
+        response = sqs_service.get_exact_amount_out_custom_direct_quote(token_out, denom_in, pool_id)
+        elapsed_time_ms = (time.time() - start_time) * 1000
+
+        assert response.status_code == expected_status_code, f"Error: {response.text}"
+        assert expected_latency_upper_bound_ms > elapsed_time_ms, f"Error: latency {elapsed_time_ms} exceeded {expected_latency_upper_bound_ms} ms, denom in {denom_in} and token out {token_out}" 
+
+        response_json = response.json()
+        
+        print(response.text)
```

Comment:
nit: I keep seeing this and similar boilerplate repeatedly. Is there a chance we could abstract it with a helper?